### PR TITLE
gRPC-c data parsing

### PR DESCRIPTION
**Pull request by orishuss**
Stirling now registers on perf buffers where the gRPC-c eBPF module writes data to.
There are 3 buffers:
1. gRPC events
2. gRPC headers
3. close events

The logic of handling gRPC sessions works for Golang events. This logic is now used for gRPC-c events as well.
The data that the gRPC-c eBPF module passes to the user-space differs from the data that the Golang gRPC eBPF module passes to the user-space.
This PR is basically an abstraction layer that "translates" gRPC-c eBPF events to the known format of Golang gRPC events.

gRPC-c events are still not enabled; They will be enabled in the next PR, where the needed probes will be attached by the UProbeManager. 
However, the gRPC-c eBPF program is now compiled; because in order for the code to find the perf buffers, they must exist.

### CODE REVIEW

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 1]
```diff
@@ -845,6 +904,158 @@ void SocketTraceConnector::AcceptHTTP2Data(std::unique_ptr<HTTP2DataEvent> event
...
+      d_event_ready->data_attr = current_data_event.data_attr;
+      // Assign instead of copy so null terminator is copied too.
+      // Assign can throw when wrong length is given, and since the data is brought from eBPF,
+      // better catch.
+      d_event_ready->payload =
+          std::string_view(data_slice.bytes, current_data_event.data_attr.data_buf_size);
+    } catch (const std::exception& e) {
+      LOG(WARNING) << absl::Substitute("gRPC-C slice data assigning threw exception $0", e.what());
+      return;
+    }
```
Comment (yzhao1012) [inline src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:952]: Pixie C++ code don't use exception, following the google's style guide.
Where exception can be thrown? It seems everything is just assignment of PoD types, which should not throw.

**Comment by yzhao1012** [comment]
The probes are perf_buffers are not attached and opened yet. So we do not need add automated tests in this PR.
As a precaution, run the following command at the top of the repository directory:
scripts/sudo_bazel_run.sh //src/stirling/source_connectors/socket_tracer:socket_trace_bpf_test 

And afterwards when Omid actually merge this PR, Pixie's internal Jenkins will verify this PR on 4.14 kernel, which is the version Pixie was committed to support at the beginning.

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 2]
```diff
@@ -360,6 +361,12 @@ auto SocketTraceConnector::InitPerfBufferSpecs() {
        PerfBufferSizeCategory::kControl},
       {"go_grpc_events", HandleHTTP2Event, HandleHTTP2EventLoss, kTargetDataBufferSize,
        PerfBufferSizeCategory::kData},
+      {"grpc_c_events", HandleGrpcCEvent, HandleGrpcCDataLoss, kTargetControlBufferSize,
+       PerfBufferSizeCategory::kData},
+      {"grpc_c_header_events", HandleGrpcCHeaderEvent, HandleGrpcCHeaderDataLoss,
+       kTargetControlBufferSize, PerfBufferSizeCategory::kData},
+      {"grpc_c_close_events", HandleGrpcCCloseEvent, HandleGrpcCCloseDataLoss,
+       kTargetControlBufferSize, PerfBufferSizeCategory::kControl},
```
[the remark this answers is not in the document]
Comment (orishuss) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:369]: Ah, I hadn't noticed there are two different sizes :)

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 3]
```diff
@@ -845,6 +904,158 @@ void SocketTraceConnector::AcceptHTTP2Data(std::unique_ptr<HTTP2DataEvent> event
...
+                                       event->conn_id.upid.pid, event->direction);
+      return;
+  }
+
+  // Prepare list of slices.
+  std::vector<struct grpc_c_data_slice_t> slices(0);
+  slices.push_back(event->slice);
+
+  this->AcceptGrpcCEvent(event->conn_id, event->stream_id, event->timestamp, outgoing,
+                         event->position_in_stream, slices);
```
[the remark this answers is not in the document]
Comment (orishuss) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:978]: Done

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 4]
```diff
@@ -845,6 +904,158 @@ void SocketTraceConnector::AcceptHTTP2Data(std::unique_ptr<HTTP2DataEvent> event
...
+  this->AcceptHTTP2Header(std::move(h_event_ready));
+}
+
+void SocketTraceConnector::AcceptGrpcCCloseEvent(
+    std::unique_ptr<struct grpc_c_stream_closed_data> event) {
+  // Initiate header event template as if it arrived from Golang gRPC eBPF.
+  struct go_grpc_http2_header_event_t header_event_data_go_style = {};
+  InitiateHeaderEventDataGoStyle(
+      event->conn_id, event->stream_id, event->timestamp, true,
+      true,  // Whether outgoing. This does not matter because it's overridden later.
```
[the remark this answers is not in the document]
Comment (orishuss) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:1037]: Done

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc (hunk 1 above)
Comment (orishuss) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:952]: Done

**Comment by orishuss** [comment]
> The probes are perf_buffers are not attached and opened yet. So we do not need add automated tests in this PR. As a precaution, run the following command at the top of the repository directory: scripts/sudo_bazel_run.sh //src/stirling/source_connectors/socket_tracer:socket_trace_bpf_test
> 
> And afterwards when Omid actually merge this PR, Pixie's internal Jenkins will verify this PR on 4.14 kernel, which is the version Pixie was committed to support at the beginning.

Well, there's one warning I'm not sure why happens (the loop looks bound to me):
`warning: ./src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c:645:3: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering`

And also the tests crash with a segfault at the beginning. When running Stirling wrapper (not within tests) it works as expected, so I don't know what they do or why they'd crash, gonna dive into it. If you have any idea due to your experience with the tests, let me know where to start. Thanks :)

**Comment by yzhao1012** [comment]
> > The probes are perf_buffers are not attached and opened yet. So we do not need add automated tests in this PR. As a precaution, run the following command at the top of the repository directory: scripts/sudo_bazel_run.sh //src/stirling/source_connectors/socket_tracer:socket_trace_bpf_test
> > And afterwards when Omid actually merge this PR, Pixie's internal Jenkins will verify this PR on 4.14 kernel, which is the version Pixie was committed to support at the beginning.
> 
> @yzhao1012 Well, there's one warning I'm not sure why happens (the loop looks bound to me): `warning: ./src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c:645:3: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering`
> 
> And also the tests crash with a segfault at the beginning. When running Stirling wrapper (not within tests) it works as expected, so I don't know what they do or why they'd crash, gonna dive into it. If you have any idea due to your experience with the tests, let me know where to start. Thanks :)

You might want to try to change the macro definition to 
#define SIZE_OF_DATA_SLICE_ARRAY 8

The extra parentheses might confuses the complier. Other than that I couldn't think of anything.

**Comment by orishuss** [comment]
Updates:
1. The cause of the segmentation fault was too many open files in the linux system. The solution is running `ulimit -nS 2048` before running the tests (the previous maximum open files for me was 1024). The reason it crashed is that bcc failed to attach probes (i got the error `create_uprobe_event: open(/sys/kernel/debug/tracing/uprobe_events) Too many open files` thrown at `https://github.com/iovisor/bcc/blob/v0.24.0/src/cc/libbpf.c#L1099`). This error does not seem to happen when I run Stirling without the tests, so I guess it's either because the tests open more files, or because the tests run in a different manner than Stirling does (may be a different container configuration?). Anyway, unless you think otherwise, it seems fine by me.
2. The parentheses are not the thing that cause the unrolling to fail. I don't know why it happens, and everything seems to run smoothly. Do you think we should keep looking for the root cause, or perhaps it's OK if everything runs fine?
3. The tests take a while to run, I will update on their result shortly.

**Review by yzhao1012 — COMMENTED** [review]
BCC initialization time went from <10s to >40s. This causes test timeout.
I think we have to resolve this before merging:
Either reduce the initialization time to close to <20s (roughly double from current time).
Or conclude that the time spent cannot be avoided.

I'll spend more time on this.

**Comment by yzhao1012** [comment]
I rewrote 2 places of grpc_c_trace.c, and now the BPF::init() takes a little over 10 seconds, down from 40+ seconds. The tests are not timing out any more.

https://github.com/pixie-io/pixie/commit/c2abc41b814a45964b94a7dde846836fc794f9ba
https://github.com/pixie-io/pixie/commit/7b9e6032b84f134a058df8f61c02e5f85374285e

Could you resync this PR?
Then we can merge it.

**Comment by avivzgroundcover** [comment]
I noticed that removing the initialization loops creates a bug with the headers parsing - since we don’t initialize the metadata map every time, the headers from previous requests remain. If we copy a shorter header into a place where a longer header was, the trailing characters of the longer header will still be there, adding themselves to the new, shorter header. I saw this occurring in our test environment, verifying the bug exists.

The solution to this is either zeroing out the header arrays before writing to them, or explicitly adding a null terminator for each header in the offset fitting for that header, so that the trailing characters are ignored when parsing as string. 
I implemented the null terminator solution and it looks ok in our environment. Should it be added to this PR, opened in a new PR, or anything else?

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 5]
```diff
@@ -22,6 +22,7 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+#include <algorithm>
```
Comment (yzhao1012) [inline src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:25]: Is <algorithm> header needed?

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 6]
```diff
@@ -801,6 +808,58 @@ void SocketTraceConnector::HandleHTTP2EventLoss(void* cb_cookie, uint64_t lost)
...
 }
 
+void SocketTraceConnector::HandleGrpcCHeaderEvent(void* cb_cookie, void* data, int /*data_size*/) {
+  DCHECK(data != nullptr) << "GrpcCHeaderEvent data argument is null.";
+  DCHECK(cb_cookie != nullptr) << "Perf buffer callback not set-up properly. Missing cb_cookie.";
+
+  auto* connector = static_cast<SocketTraceConnector*>(cb_cookie);
+  struct grpc_c_header_event_data_t* event_data =
+      static_cast<struct grpc_c_header_event_data_t*>(data);
+  auto event = std::make_unique<struct grpc_c_header_event_data_t>(*event_data);
```
[the remark this answers is not in the document]
Comment (yzhao1012) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:822]: If time is constrained, then please leave a TODO(yzhao): <the above suggestion> and let me followup.

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (orishuss) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc]: You are correct

File: src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/grpc_c.h [hunk 7]
```diff
@@ -18,6 +18,8 @@
 
 #pragma once
 
+#include "src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/common.h"
```
[the remark this answers is not in the document]
Comment (orishuss) [reply src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/grpc_c.h:21]: It is needed for the traffic_direction enum and the conn_id struct.

**Comment by oazizi000** [comment]
Merged in 0fc4472451c020e217026811169e5fb3ec91ee69

**Pull request closed without merging**